src/server/dflycmd.cc

```diff
@@ -70,8 +95,22 @@ void DflyCmd::Run(CmdArgList args, ConnectionContext* cntx) {
 }
 
 void DflyCmd::OnClose(ConnectionContext* cntx) {
-  if (cntx->conn_state.repl_session_id > 0 && cntx->conn_state.repl_threadid != kuint32max) {
-    DeleteSyncSession(cntx->conn_state.repl_session_id);
+  unsigned session_id = cntx->conn_state.repl_session_id;
+  unsigned flow_id = cntx->conn_state.repl_flow_id;
+
+  if (!session_id)
+    return;
+
+  if (flow_id == kuint32max) {
+    DeleteSyncSession(session_id);
+  } else {
+    shared_ptr<SyncInfo> sync_info = GetSyncInfo(session_id);
+    if (sync_info) {
+      lock_guard lk(sync_info->mu);
+      if (sync_info->state != SyncState::CANCELLED) {
+        UnregisterFlow(&sync_info->flows[flow_id]);
+      }
```

Comment:
Explicitly checking for the `CANCELLED` state is probably a better solution than letting `DeleteSyncSession` acquire the sync info mutex for full duration. This way blocking is much shorter